File: 05 - Adapter/Python/electric_socket.py
class EuropeanSocketInterface:
    def voltage(self):
        pass

    def live(self):
        pass

    def neutral(self):
        pass

    def earth(self):
        pass


class Socket(EuropeanSocketInterface):
    def voltage(self):
        return 230

    def live(self):
        return 1

    def neutral(self):
        return -1

    def earth(self):
        return 0


class USASocketInterface:
    def voltage(self):
        pass

    def live(self):
        pass

    def neutral(self):
        pass


class Adapter(USASocketInterface):
    def __init__(self, socket):
        self.socket = socket

    def voltage(self):
        return 110

    def live(self):
        return self.socket.live()

    def neutral(self):
        return self.socket.neutral()


class ElectricKettle:
    def __init__(self, power):
        self.power = power

    def boil(self):
        if self.power.voltage() > 110:
            print("Kettle on fire!")
        else:
            if self.power.live() == 1 and self.power.neutral() == -1:
                print("Coffee time!")
            else:
                print("No power.")
                
if __name__ == "__main__":
    socket = Socket()
    adapter = Adapter(socket)
    kettle = ElectricKettle(adapter)

    kettle.boil()

File: 05 - Adapter/Dart/electric_socket.dart
void main(){
  Socket socket = Socket();
  Adapter adapter = Adapter(socket);
  USElectricKettle kettle = USElectricKettle(adapter);
  kettle.boil();
}

abstract class EuropeanSocketInterface {
  int voltage();
  int live();
  int neutral();
  int earth();
}

abstract class USASocketInterface {
  int voltage();
  int live();
  int neutral();
}

class Socket implements EuropeanSocketInterface {
  @override
  int voltage() => 230;
  
  @override
  int live() => 1;
  
  @override
  int neutral() => -1;
  
  @override
  int earth() => 0;
}

class Adapter implements USASocketInterface {
  final EuropeanSocketInterface _socket;
  
  Adapter(this._socket);
  
  @override
  int voltage() => 110;
  
  @override
  int live() => _socket.live();
  
  @override
  int neutral() => _socket.neutral();
}

class USElectricKettle {
  final USASocketInterface _power;
  
  USElectricKettle(this._power);
  
  void boil(){
    if (_power.voltage() > 110) {
      print("Kettle on fire!");
    }
    else if (_power.live() == 1 && _power.neutral() == -1){
      print("Coffee time!");
    }
    else {
      print("No power.");
    }
  }
}